zypp-plugin/snapper_zypp_plugin.cc

```diff
@@ -0,0 +1,341 @@
+    bool seen_first = false;
+    os << '{';
+    for (auto s: ss) {
+	if (seen_first) {
+	    os << ", ";
+	}
+	os << s;
+	seen_first = true;
+    }
+    os << '}';
+    return os;
+}
+
+// Normally the only configuration this program needs is
+// the /etc/snapper/zypp-plugin.conf file.
+// But for testing we need more places to inject mocks.
+// This is done with SNAPPER_ZYPP_PLUGIN_* environment variables.
+// (Using argv is not useful since libzypp does not use it in the
+// plugin protocol.)
+class ProgramOptions {
+public:
+    string plugin_config;
+    string snapper_config;
+    DBusBusType bus;
+
+    ProgramOptions()
+    : plugin_config("/etc/snapper/zypp-plugin.conf")
+    , snapper_config("root")
+    , bus(DBUS_BUS_SYSTEM)
+    {
+	const char * s;
+
+	s = getenv("SNAPPER_ZYPP_PLUGIN_CONFIG");
+	if (s != nullptr) {
+	    plugin_config = s;
+	}
+
+	s = getenv("SNAPPER_ZYPP_PLUGIN_SNAPPER_CONFIG");
+	if (s != nullptr) {
+	    snapper_config = s;
+	}
+
+	s = getenv("SNAPPER_ZYPP_PLUGIN_DBUS_SESSION");
+	if (s != nullptr) {
+            bus = DBUS_BUS_SESSION;
+        }
+    }
+};
+
+class SnapperZyppPlugin : public ZyppCommitPlugin {
+public:
+    SnapperZyppPlugin(const ProgramOptions& opts)
+    : snapper_cfg(opts.snapper_config)
+    , dbus_conn(opts.bus)
+    , pre_snapshot_num(0)
+    , solvable_matchers(SolvableMatcher::load_config(opts.plugin_config))
+    {
+	string caller_prog;
+	readlink(sformat("/proc/%d/exe", getppid()), caller_prog);
+	snapshot_description = sformat("zypp(%s)", basename(caller_prog).c_str());
+    }
+
+    Message plugin_begin(const Message& m) override {
+	cerr << "INFO:" << "PLUGINBEGIN" << endl;
+	userdata = get_userdata(m);
+
+	return ack();
+    }
+    Message plugin_end(const Message& m) override {
+	cerr << "INFO:" << "PLUGINEND" << endl;
+	return ack();
+    }
+    Message commit_begin(const Message& msg) override {
+	cerr << "INFO:" << "COMMITBEGIN" << endl;
+
+	set<string> solvables = get_solvables(msg, Phase::BEFORE);
+	cerr << "DEBUG:" << "solvables: " << solvables << endl;
+
+	bool found, important;
+	match_solvables(solvables, found, important);
+	cerr << "INFO:" << "found: " << found << ", important: " << important << endl;
+
+	if (found || important) {
+	    userdata["important"] = important ? "yes" : "no";
+
+	    try {
+		cerr << "INFO:" << "creating pre snapshot" << endl;
+		pre_snapshot_num = command_create_pre_snapshot(
+                    dbus_conn, snapper_cfg,
+                    snapshot_description, cleanup_algorithm, userdata
+                );
+		cerr << "DEBUG:" << "created pre snapshot " << pre_snapshot_num << endl;
+	    }
+            catch (const DBus::ErrorException& ex) {
+		SN_CAUGHT(ex);
+		cerr << "ERROR:" << error_description(ex) << endl;
+	    }
+	    catch (const Exception& ex) {
+		SN_CAUGHT(ex);
+	    }
+	}
+
+	return ack();
+    }
+
+    Message commit_end(const Message& msg) override {
+	cerr << "INFO:" << "COMMITEND" << endl;
```

Comment:
NP: why `cerr` ? I would expect to use `clog` here.

Reply:
1. I forgot about [`clog`](http://www.cplusplus.com/reference/iostream/clog/). Duh.
2. Since the difference is that `clog` is a plain stream whereas `cerr` makes `cout` flush and also flushes itself, I think `cerr` is better for stdio-based interprocess communication: in case this plugin crashes, we want libzypp to have received the recent log messages.